### ISSUE TITLE
Text-based matching of exclusions can catch in other meta-packages that put exclusion marks in strings

### ISSUE THREAD

**Issue by MichaelChirico** [r-lib/lintr#1990]
Split off from #1987.

Here's a minimal example:

```r
lintr::lint(linters = lintr::assignment_linter(), "
a = 1 # nolint: assignment_linter.
'# nolint: %s.'
")
Warning message:
In add_exclusions(exclusions, nolints[i], linters_string, exclude_linter_sep,  :
  Could not find linter named ‘%s’ in the list of active linters. Make sure the linter is uniquely identified by the given n
```

**Comment by AshesITR** [comment]
If we want to avoid the overhead of parsing each file twice, we would need to pull apart the line exclusion logic from the file exclusion logic. It does seem worth it in my opinion, because the current implementation could also lead to accidental nolints.

**Comment by MichaelChirico** [comment]
Authors affected by this should have a look at #2160. I don't think there's much else we can do.
